File: UCF HSPT Documents/2013/Solutions/cosmic.c

// 27th Annual UCF High School Programming Tournament
//
// Problem Name: Greatest Cosmic Devourers (cosmic)
//
// Author:        Mike Galletti
// Judge Data:    Matt Fontaine
// C Solution:    Aaron Teetor
// Java Solution: Gabe Pita
// Verifier:      Andrew Harn

#include <stdio.h>
#include <stdlib.h>

int main(void) {
   //Open the file
   FILE* fp;
   fp = fopen("cosmic.in", "r");

   int numScenarios;
   int i, j, k;

   //Read in the number of scenarios (cases)
   fscanf(fp, "%d", &numScenarios);

   //Loop over the scenarios
   for (i = 0; i < numScenarios; i++)
   {
      //Read in a and b
      int a;
      int b;
      fscanf(fp, "%d", &a);
      fscanf(fp, "%d", &b);

      //Planet 0 always survives, start on planet 1
      int planetsSurviving = 1;
      int planetsChecked = 1;

      //Loop through all planets until we get to a contested one.
      //Note that the sequence repeats after that so enough planets 
      //will have been checked once we get to a contested one!
      while (planetsChecked % a != 0 || planetsChecked % b != 0) {
         //If a nor b wants to eat the planet it counts as surviving
         if (planetsChecked % a != 0 && planetsChecked % b != 0) {
            planetsSurviving++;
         }

         //Add one to number of planets checked
         planetsChecked++;
      }

      //Take the percent of planets that survived
      printf("Universe #%d: There's a %.4f%% chance we'll survive!\n", i + 1,
             100 * ((double)(planetsSurviving) / (double)(planetsChecked)));
   }

   //Exit
   fclose(fp);
   return EXIT_SUCCESS;
}

File: UCF HSPT Project/src/y2013/cosmic.java

// 27th Annual UCF High School Programming Tournament
//
// Problem Name: Greatest Cosmic Devourers (cosmic)
//
// Author:        Mike Galletti
// Judge Data:    Matt Fontaine
// C Solution:    Aaron Teetor
// Java Solution: Gabe Pita
// Verifier:      Andrew Harn

import java.util.Scanner;
import java.io.File;
import java.io.IOException;


public class cosmic {

   public static void main(String[] args) throws IOException
   {
      //Open file for reading
      Scanner in = new Scanner(new File("cosmic.in"));
      
      //The number of test cases
      int u = in.nextInt();
      
      //Loop over the test cases
      for(int i =1; i <= u; i++)
      {
         //Read in the factors for each monster
         int a = in.nextInt();
         int b = in.nextInt();
         
         //The pattern of stars being devoured repeats every time a star 
         //is divisible by both a and b.  All of these stars are multiples 
         //of the least common denominator of a and b, but all you need to know
         //to solve this question is that a times b is divisible by 
         //both a and b.
         
         //We get the percentage of stars that are devoured from 
         //0 to a times b, and that will be the percentage for the 
         //universe since that pattern found in 0 to a times b will 
         //repeat endlessly
         
         int ate = 0;
         
         //Iterate through the stars till we get to the repeating pattern
         for(int j =0; j < a * b; j++)
         {
            //If the star isn't contested by both monsters
            if(j % a != 0 || j % b !=0)
            {
               //And one of them wishes to eat it, then increment 
               //the number of stars eaten
               if(j % a == 0)
                  ate++;
               if(j % b == 0)
                  ate++;
            }
         }
         
         //The percentage of stars ate is simply the number eaten 
         //over the number of stars considered
         double percentAte = (ate / (double) (a * b)) * 100; 
         
         //The survival percentage is the whole minus the percentage 
         //of eaten stars
         double percentSurvived = 100 - percentAte;
         
         //Print out the answer
         System.out.printf("Universe #%d: There's a %.4f",i,percentSurvived);
         System.out.println("% chance we'll survive!");
      }

      //Done
      in.close();
   }
}
